File: linux/test/tls/sslserver.cpp
#include <iostream>
#include <cstring>
#include <unistd.h>
#include <arpa/inet.h>
#include <openssl/ssl.h>
#include <openssl/bio.h>
#include <openssl/err.h>

#define CERT_FILE "server.crt"
#define KEY_FILE "server.key"

void handle_client(SSL* ssl) {
    char buffer[1024] = {0};
    int bytes_read = SSL_read(ssl, buffer, sizeof(buffer));
    if (bytes_read > 0) {
        std::cout << "Received request: " << buffer << std::endl;
        const char* response = "HTTP/1.1 200 OK\r\nContent-Length: 13\r\n\r\nHello, world!";
        SSL_write(ssl, response, strlen(response));
    }
    SSL_shutdown(ssl);
    SSL_free(ssl);
}

int main() {
    SSL_load_error_strings();
    SSL_library_init();
    OpenSSL_add_all_algorithms();

    const char* cert_file = CERT_FILE;
    const char* key_file = KEY_FILE;

    SSL_CTX* ctx = SSL_CTX_new(TLS_server_method());
    if (!ctx) {
        std::cerr << "Error creating SSL context." << std::endl;
        return 1;
    }

    if (SSL_CTX_use_certificate_file(ctx, cert_file, SSL_FILETYPE_PEM) <= 0 ||
        SSL_CTX_use_PrivateKey_file(ctx, key_file, SSL_FILETYPE_PEM) <= 0) {
        std::cerr << "Error loading certificate or private key." << std::endl;
        return 1;
    }

    int server_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket < 0) {
        std::cerr << "Error creating socket." << std::endl;
        return 1;
    }

    struct sockaddr_in server_addr;
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(12345);
    server_addr.sin_addr.s_addr = INADDR_ANY;

    if (bind(server_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        std::cerr << "Error binding socket." << std::endl;
        return 1;
    }

    if (listen(server_socket, 5) < 0) {
        std::cerr << "Error listening on socket." << std::endl;
        return 1;
    }

    std::cout << "Server is listening on 127.0.0.1:12345..." << std::endl;

    while (true) {
        struct sockaddr_in client_addr;
        socklen_t client_addr_len = sizeof(client_addr);
        int client_socket = accept(server_socket, (struct sockaddr*)&client_addr, &client_addr_len);
        if (client_socket < 0) {
            std::cerr << "Error accepting connection." << std::endl;
            continue;
        }

        SSL* ssl = SSL_new(ctx);
        if (!ssl) {
            std::cerr << "Error creating SSL object." << std::endl;
            close(client_socket);
            continue;
        }

        SSL_set_fd(ssl, client_socket);
        if (SSL_accept(ssl) <= 0) {
            std::cerr << "Error performing SSL handshake." << std::endl;
            SSL_free(ssl);
            close(client_socket);
            continue;
        }

        std::cout << "Accepted connection from " << inet_ntoa(client_addr.sin_addr)
                  << ":" << ntohs(client_addr.sin_port) << std::endl;

        handle_client(ssl);
    }

    SSL_CTX_free(ctx);
    close(server_socket);
    return 0;
}

File: linux/test/tls/sslserver.py
import socket
import ssl

def create_ssl_context(certfile, keyfile):
    context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
    context.load_cert_chain(certfile, keyfile)
    return context

def handle_client(client_socket):
    request = client_socket.recv(1024)
    print(f"Received request: {request.decode()}")

    response = b"HTTP/1.1 200 OK\r\nContent-Length: 13\r\n\r\nHello, world!"
    client_socket.sendall(response)
    client_socket.close()

def main():
    HOST = '127.0.0.1'
    PORT = 12345
    CERTFILE = 'server.crt'
    KEYFILE = 'server.key'

    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server_socket.bind((HOST, PORT))
    server_socket.listen(5)

    ssl_context = create_ssl_context(CERTFILE, KEYFILE)

    print(f"Server is listening on {HOST}:{PORT}...")
    while True:
        client_socket, client_addr = server_socket.accept()
        ssl_socket = ssl_context.wrap_socket(client_socket, server_side=True)

        print(f"Accepted connection from {client_addr[0]}:{client_addr[1]}")
        handle_client(ssl_socket)

if __name__ == "__main__":
    main()
